### PR TITLE
hw-03

### PR DESCRIPTION
**Pull request by niksaz**

### CODE REVIEW

File: 3/classification/features.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+from sklearn.preprocessing import scale
+
+from kaldi.feat.mfcc import Mfcc, MfccOptions
+from kaldi.matrix import SubVector, SubMatrix
+from kaldi.util.table import SequentialWaveReader, SequentialMatrixReader, MatrixWriter
+
+
+def compute_kaldi_features(desc_path, output_path, samp_freq=8000):
+    print('Started processing', desc_path)
+    rspec, wspec = 'scp:{}'.format(desc_path), 'ark:{}'.format(output_path)
```
Comment (Rauf-Kurbanov) [inline 3/classification/features.py]: Consider f-string https://www.python.org/dev/peps/pep-0498/

File: 3/classification/training.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    N_train = len(train_features)
+    N_test = len(test_features)
+    train_ids = list(range(N_train))
+    test_ids = list(range(N_test))
+
+    train_losses = []
+    test_accuracies = []
+
+    for _ in tqdm(range(EPOCHS)):
```
Comment (Rauf-Kurbanov) [inline 3/classification/training.py:67]: Consider `trange`

File: 3/classification/training.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        np.random.shuffle(train_ids)
+
+        model.train()
+        train_loss = 0
+        for i in range(0, N_train, BATCH_SIZE):
+            model.zero_grad()
+            train_inputs, train_lens, train_targets = pack_feature_class_lists(
+                train_features, train_classes, train_ids[i: i +BATCH_SIZE], device)
```
Comment (Rauf-Kurbanov) [inline 3/classification/training.py]: seems like you could have just wrapped this into pytorch Dataloader

File: 3/classification/training.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        train_losses.append(train_loss)
+
+        model.eval()
+        correct = 0
+        for i in range(0, N_test, BATCH_SIZE):
+            test_inputs, test_lens, test_targets = pack_feature_class_lists(
+                test_features, test_classes ,test_ids[i: i +BATCH_SIZE], device)
+            test_predictions = model.predict_classes(test_inputs, test_lens).cpu()
+            correct += (test_predictions == test_targets).sum().item()
+        accuracy = 100 * correct / N_test
```
Comment (Rauf-Kurbanov) [inline 3/classification/training.py]: In future I would suggest using some libraries for metrics like sklearn because this way it is easier to add metrics and also does not require to deal with edgecased by yourself (as division by zero)

File: 3/classification/training.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            X_batch = torch.tensor(seq, dtype=torch.float).to(device)
+            Y_pred = model(X_batch).cpu()
+            seq = list(map(lambda ind: ys[ind], ids))
+            Y_batch = torch.tensor(seq, dtype=torch.float)
+            loss = criterion(Y_pred, Y_batch)
+            loss.backward()
+            optimizer.step()
+            train_loss += loss.item()
+        train_losses.append(train_loss)
+    model.eval()
```
Comment (Rauf-Kurbanov) [inline 3/classification/training.py]: it may be unexpected by your class user to get model already in eval mode

**Review by Rauf-Kurbanov — APPROVED** [review]
Work seems pretty nice, one thing I can say for sure you could train your model for a little longer
Great visualisation thought. 5/4

File: 3/classification/features.py (hunk 1 above)
Comment (niksaz) [reply 3/classification/features.py]: Updated.

File: 3/classification/training.py (hunk 2 above)
Comment (niksaz) [reply 3/classification/training.py:67]: It is good to know about it. But tqdm does not work well with jupyter use tqdm_notebook.

File: 3/classification/training.py (hunk 3 above)
Comment (niksaz) [reply 3/classification/training.py]: Great suggestion!

File: 3/classification/training.py (hunk 4 above)
Comment (niksaz) [reply 3/classification/training.py]: Right, it is just not easy to use those metrics when using batches.

File: 3/classification/training.py (hunk 5 above)
Comment (niksaz) [reply 3/classification/training.py]: Updated.

**Comment by niksaz** [comment]
Thank you, @Rauf-Kurbanov, for the comments!
